pdns/ednscookies.cc

```diff
@@ -19,33 +19,158 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 #include "ednscookies.hh"
+#include "misc.hh"
+#include "string_compare.hh"
 
-bool getEDNSCookiesOptFromString(const string& option, EDNSCookiesOpt* eco)
+#ifdef HAVE_CRYPTO_SHORTHASH
+#include <sodium.h>
+#endif
+
+EDNSCookiesOpt::EDNSCookiesOpt(const std::string& option)
 {
-  return getEDNSCookiesOptFromString(option.c_str(), option.length(), eco);
+  getEDNSCookiesOptFromString(option.c_str(), option.length());
 }
 
-bool getEDNSCookiesOptFromString(const char* option, unsigned int len, EDNSCookiesOpt* eco)
+EDNSCookiesOpt::EDNSCookiesOpt(const char* option, unsigned int len)
 {
-  if(len != 8 && len < 16)
-    return false;
-  eco->client = string(option, 8);
-  if (len > 8) {
-    eco->server = string(option + 8, len - 8);
-  }
-  return true;
+  getEDNSCookiesOptFromString(option, len);
+}
+
+bool EDNSCookiesOpt::makeFromString(const std::string& option)
+{
+  getEDNSCookiesOptFromString(option.c_str(), option.length());
+  return isWellFormed();
+}
+
+bool EDNSCookiesOpt::makeFromString(const char* option, unsigned int len)
+{
+  getEDNSCookiesOptFromString(option, len);
+  return isWellFormed();
 }
 
-string makeEDNSCookiesOptString(const EDNSCookiesOpt& eco)
+string EDNSCookiesOpt::makeOptString() const
 {
   string ret;
-  if (eco.client.length() != 8)
-    return ret;
-  if (eco.server.length() != 0 && (eco.server.length() < 8 || eco.server.length() > 32))
+  if (!isWellFormed())
     return ret;
-  ret.assign(eco.client);
-  if (eco.server.length() != 0)
-    ret.append(eco.server);
+  ret.assign(client);
+  if (server.length() != 0)
+    ret.append(server);
   return ret;
 }
+
+void EDNSCookiesOpt::getEDNSCookiesOptFromString(const char* option, unsigned int len)
+{
+  client.clear();
+  server.clear();
+  if (len < 8)
+    return;
+  client = string(option, 8);
+  if (len > 8) {
+    server = string(option + 8, len - 8);
+  }
+}
+
+bool EDNSCookiesOpt::isValid(const string& secret, const ComboAddress& source)
+{
+#ifdef HAVE_CRYPTO_SHORTHASH
+  if (server.length() != 16 || client.length() != 8) {
+    return false;
+  }
+  if (server[0] != '\x01') {
+    // Version is not 1, can't verify
+    return false;
+  }
+  uint32_t ts;
+  memcpy(&ts, &server[4], sizeof(ts));
+  ts = ntohl(ts);
+  uint32_t now = static_cast<uint32_t>(time(nullptr));
+  // RFC 9018 section 4.3:
+  //    The DNS server
+  //    SHOULD allow cookies within a 1-hour period in the past and a
+  //    5-minute period into the future
+  if (rfc1982LessThan(now + 300, ts) && rfc1982LessThan(ts + 3600, now)) {
+    return false;
+  }
+  if (secret.length() != crypto_shorthash_KEYBYTES) {
+    // XXX should we throw std::range_error here?
```

Comment:
better to check it at startup, where you already have a catch for range_error anyway

Reply:
we do, but this function could potentially also be called from other places to do cookie things

Reply:
We only check the length of the input string at startup - but I just re-read `makeBytesFromHex` and clearly it cannot generate an output string of the wrong length.

So, I guess the comment can just go.